// node-sp-auth/lib/src/utils/AdfsHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request = require("request-promise");
const url = require("url");
const _ = require("lodash");
const fs = require("fs");
const path = require("path");
let xmldoc = require('xmldoc');
class AdfsHelper {
    static getSamlAssertion(siteUrl, credentials) {
        let adfsHost = url.parse(credentials.adfsUrl).host;
        let usernameMixedUrl = `https://${adfsHost}/adfs/services/trust/13/usernamemixed`;
        let samlTemplate = fs.readFileSync(path.join(__dirname, '..', '..', 'templates', 'adfs_saml_wsfed.tmpl'));
        let samlBody = _.template(samlTemplate.toString())({
            to: usernameMixedUrl,
            username: credentials.username,
            password: credentials.password,
            relyingParty: credentials.relyingParty
        });
        return request.post(usernameMixedUrl, {
            body: samlBody,
            strictSSL: false,
            simple: false,
            headers: {
                'Content-Length': samlBody.length,
                'Content-Type': 'application/soap+xml; charset=utf-8'
            }
        })
            .then(xmlResponse => {
            let doc = new xmldoc.XmlDocument(xmlResponse);
            let tokenResponseCollection = doc.childNamed('s:Body').firstChild;
            if (tokenResponseCollection.name.indexOf('Fault') !== -1) {
                throw new Error(tokenResponseCollection.toString());
            }
            let responseNamespace = tokenResponseCollection.name.split(':')[0];
            let securityTokenResponse = doc.childNamed('s:Body').firstChild.firstChild;
            let samlAssertion = securityTokenResponse.childNamed(responseNamespace + ':RequestedSecurityToken').firstChild;
            let notBefore = samlAssertion.firstChild.attr['NotBefore'];
            let notAfter = samlAssertion.firstChild.attr['NotOnOrAfter'];
            return {
                value: samlAssertion.toString({ compressed: true }),
                notAfter: notAfter,
                notBefore: notBefore
            };
        });
    }
}
exports.AdfsHelper = AdfsHelper;
