// node-sp-auth/lib/src/auth/IAuthOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url = require("url");
function isAddinOnlyOnline(T) {
    return T.clientSecret !== undefined;
}
exports.isAddinOnlyOnline = isAddinOnlyOnline;
function isAddinOnlyOnpremise(T) {
    return T.shaThumbprint !== undefined;
}
exports.isAddinOnlyOnpremise = isAddinOnlyOnpremise;
function isUserCredentialsOnline(siteUrl, T) {
    let host = (url.parse(siteUrl)).host;
    let isOnPrem = host.indexOf('.sharepoint.com') === -1 && host.indexOf('.sharepoint.cn') === -1;
    if (!isOnPrem && T.username !== undefined && !isAdfsCredentials(T)) {
        return true;
    }
    return false;
}
exports.isUserCredentialsOnline = isUserCredentialsOnline;
function isUserCredentialsOnpremise(siteUrl, T) {
    let host = (url.parse(siteUrl)).host;
    let isOnPrem = host.indexOf('.sharepoint.com') === -1 && host.indexOf('.sharepoint.cn') === -1;
    if (isOnPrem && T.username !== undefined && !isAdfsCredentials(T)) {
        return true;
    }
    return false;
}
exports.isUserCredentialsOnpremise = isUserCredentialsOnpremise;
function isTmgCredentialsOnpremise(siteUrl, T) {
    let host = (url.parse(siteUrl)).host;
    let isOnPrem = host.indexOf('.sharepoint.com') === -1 && host.indexOf('.sharepoint.cn') === -1;
    if (isOnPrem && T.username !== undefined && T.tmg) {
        return true;
    }
    return false;
}
exports.isTmgCredentialsOnpremise = isTmgCredentialsOnpremise;
function isFbaCredentialsOnpremise(siteUrl, T) {
    let host = (url.parse(siteUrl)).host;
    let isOnPrem = host.indexOf('.sharepoint.com') === -1 && host.indexOf('.sharepoint.cn') === -1;
    if (isOnPrem && T.username !== undefined && T.fba) {
        return true;
    }
    return false;
}
exports.isFbaCredentialsOnpremise = isFbaCredentialsOnpremise;
function isAdfsCredentials(T) {
    return T.adfsUrl !== undefined;
}
exports.isAdfsCredentials = isAdfsCredentials;
function isOndemandCredentials(T) {
    return T.ondemand !== undefined;
}
exports.isOndemandCredentials = isOndemandCredentials;
